spec/Classes.tex

```diff
@@ -663,73 +600,453 @@ \subsection{User-Defined Constructors}
+side-effects, the language requires that all fields be initialized in field
+declaration order. This applies regardless of whether field initializations
+are omitted from the initializer body. If fields are initialized out of order,
+a compile-time error will be issued.
+
+\begin{chapelexample}{fieldsOutOfOrder.chpl}
+In the following code:
+\begin{chapel}
+class Point3D {
+  var x = 1.0;
+  var y = 1.0;
+  var z = 1.0;
+
+  proc init(x: real) {
+    this.x = x;
+    // compiler inserts "this.y = 1.0;"
+    this.z = y * 2.0;
+  }
+
+  proc init(x: real, y: real, z: real) {
+    this.x = x;
+    // Without in-order rule: compiler inserts "this.y = 1.0;"
+    this.z = z;
+    // Without in-order rule: assignment, not initialization!
+    this.y = y;
+  }
+}
+
+var A = new Point3D(1.0);
+var B = new Point3D(1.0, 2.0, 3.0);
+\end{chapel}
+\begin{chapelcompopts}
+--no-warnings
+\end{chapelcompopts}
+\begin{chapeloutput}
+fieldsOutOfOrder.chpl:12: In initializer:
+fieldsOutOfOrder.chpl:17: error: Field "y" initialized out of order
+fieldsOutOfOrder.chpl:17: note: initialization of fields before .init() call must be in field declaration order
+\end{chapeloutput}
+
+The first initializer leverages the well-defined order of omitted field
+initialization to use the default value of field \chpl{y} in order to
+explicitly initialize field \chpl{z}.
+
+The second initializer initializes field \chpl{z} before field \chpl{y},
+causing a compile-time error to be issued. Without this rule the explicit
+initialization of field \chpl{y} would be assignment, which could be confusing
+and surprising (especially for more complicated types than \chpl{real}).
+
+\end{chapelexample}
+
+\subsubsection{The complete Method}
+\label{The_Complete_Method}
+\index{classes!initializers!user-defined!complete}
+
+As the initializer makes progress, the class instance is incrementally
+initialized. In order to prevent usage of uninitialized memory, there are
+restrictions on usage of the class instance before it is fully initialized:
+
+\begin{itemize}
+\item Methods may not be invoked on partially-initialized instances
+\item \chpl{this} may not be passed to functions while partially-initialized
+\end{itemize}
+
+These rules allow all methods and functions to assume that class instances
+have been initialized, provided their value is not \chpl{nil}.
+
+\begin{rationale}
+  The compiler could conceivably attempt to analyze methods and functions to
+  determine which fields are used, and selectively allow method calls on
+  partially-initialized class instances. A simpler rule, which is easier to
+  remember and implement, is to forbid method calls on partially-initialized
+  instances.
+\end{rationale}
+
+Once the class is initialized, methods may be called and \chpl{this} may be
+passed to functions. Due to the support for omitted field initialization, there
+is potential for confusion regarding the overall status of initialization.
```

Comment:
Here again there seems to be more rationale than I'd personally want, unless it's in a rationale section.

e.g.

> Methods may be called and this may be passed to functions only after the complete method is invoked.
> Rationale: Due to the support for omitted field initialization, there is potential for confusion regarding the overall status of initialization. This confusion is addressed in the design by requiring \chpl{complete} to mark the transition between partially initialized fields and fully initialized.

Reply:
I'm not sure where it goes in here, but I keep expecting to see something about the run-time type of `this`.

Reply:
You'll see something about the runtime type in the upcoming section on inheriting classes.